src/realm/sync/instruction_applier.cpp

```diff
@@ -702,56 +726,81 @@ void InstructionApplier::operator()(const Instruction::ArrayMove& instr)
             }
             list.move(index, instr.ndx_2);
         },
-        [&](auto&&...) {
-            bad_transaction_log("Invalid path for ArrayMode");
+        [&](LstBase&) {
+            bad_transaction_log("Invalid path for ArrayMove (list)");
         },
-    };
+        [&](SetBase&) {
+            bad_transaction_log("Invalid path for ArrayMove (set)");
+        },
+        [&](Dictionary&) {
+            bad_transaction_log("Invalid path for ArrayMove (dictionary)");
+        },
+        [&](Dictionary&, Mixed) {
+            bad_transaction_log("Invalid path for ArrayMove (dictionary, key)");
+        },
+        [&](Obj&, ColKey) {
+            bad_transaction_log("Invalid path for ArrayMove (obj, col)");
+        }};
     resolve_path(instr, "ArrayMove", std::move(callback));
 }
 
 void InstructionApplier::operator()(const Instruction::ArrayErase& instr)
 {
-    auto callback = util::overload{
-        [&](LstBase& list, size_t index) {
-            if (index >= instr.prior_size) {
-                bad_transaction_log("ArrayErase: Invalid index (index = %1, prior_size = %2)", index,
-                                    instr.prior_size);
-            }
-            if (index >= list.size()) {
-                bad_transaction_log("ArrayErase: Index out of bounds (%1 >= %2)", index, list.size());
-            }
-            if (instr.prior_size != list.size()) {
-                bad_transaction_log("ArrayErase: Invalid prior_size (list size = %1, prior_size = %2)", list.size(),
-                                    instr.prior_size);
-            }
-
-            list.remove(index, index + 1);
-        },
-        [&](auto&&...) {
-            bad_transaction_log("Invalid path for ArrayErase");
-        },
-    };
-    resolve_path(instr, "ArrayErase", callback);
+    resolve_path(
```

Comment:
I think we should use the same pattern everywhere. Any particular reason not to have a `callback` variable here?